src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Hierarchical.hpp

```diff
@@ -0,0 +1,969 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <boost/functional/hash.hpp>  // IWYU pragma: keep
+#include <cmath>
+#include <cstddef>
+#include <limits>
+#include <ostream>
+#include <pup.h>  // IWYU pragma: keep
+#include <string>
+#include <type_traits>
+#include <unordered_map>
+#include <utility>  // for pair
+#include <vector>
+
+#include "DataStructures/DataBox/DataBoxTag.hpp"
+#include "DataStructures/DataVector.hpp"  // IWYU pragma: keep
+#include "DataStructures/Index.hpp"
+#include "DataStructures/ModalVector.hpp"           // IWYU pragma: keep
+#include "DataStructures/Tensor/Metafunctions.hpp"  // IWYU pragma: keep
+#include "DataStructures/Tensor/Tensor.hpp"         // IWYU pragma: keep
+#include "DataStructures/Variables.hpp"             // IWYU pragma: keep
+#include "DataStructures/VariablesHelpers.hpp"
+#include "Domain/Direction.hpp"
+#include "Domain/Element.hpp"    // IWYU pragma: keep
+#include "Domain/ElementId.hpp"  // IWYU pragma: keep
+#include "Domain/Mesh.hpp"
+#include "Domain/OrientationMap.hpp"  // IWYU pragma: keep
+#include "Domain/Tags.hpp"            // IWYU pragma: keep
+#include "ErrorHandling/Assert.hpp"
+#include "ErrorHandling/Error.hpp"
+#include "NumericalAlgorithms/LinearOperators/CoefficientTransforms.hpp"
+#include "NumericalAlgorithms/Spectral/Spectral.hpp"
+#include "Options/Options.hpp"
+#include "Utilities/Algorithm.hpp"
+#include "Utilities/ForceInline.hpp"
+#include "Utilities/Gsl.hpp"
+#include "Utilities/Math.hpp"
+#include "Utilities/TMPL.hpp"
+
+// IWYU pragma: no_include <algorithm>
+// IWYU pragma: no_forward_declare Variables
+
+/// \cond
+namespace SlopeLimiters {
+template <size_t VolumeDim, typename TagsToLimit>
+class Hierarchical;
+}  // namespace SlopeLimiters
+/// \endcond
+
+namespace SlopeLimiters {
+enum class HierarchicalType { Krivodonova };
+
+std::ostream& operator<<(std::ostream& os, HierarchicalType t) noexcept;
+
+/// Given the `NodalTag` swap the `DataVector` with a `ModalVector`.
```

Comment:
I'm not exactly sure where this tag should go. It's sort of a prefix tag, so that might be the place to put it.

Reply:
In Minmod I have the same kind of tag, except I swap the DataVector for a double... maybe we need to start collecting them.

Reply:
Hmm, yea, that's a good idea. Probably a special group of prefix tags that just swap the type

Reply:
Yes, I can factor this out. Probably the most reasonable place is `DataStructures/Tags.hpp`